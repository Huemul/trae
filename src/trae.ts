import merge from 'lodash/merge';

import createRequestBody from '../lib/create-request-body'
import createResponse from '../lib/create-response';
import { format as formatUrl } from '../lib/url';
import { TraeSettings, InstanceConfig } from './types';

const defaults: RequestInit = {
  headers: { 'Content-Type': 'application/json' },
};

function createTrae(providedConf?: Partial<TraeSettings>) {
  const config: TraeSettings = Object.freeze(merge(defaults, {
    before: (conf: RequestInit) => conf,
    onResolve: (item: unknown) => Promise.resolve(item),
    onReject: (err: unknown) => Promise.reject(err),
    ...providedConf,
  }));

  function request(endpoint: string, settings: TraeSettings) {
    // TODO: We should extract some attributes to avoid exposing unnecessary
    //       data to the before middleware and fetch function.
    //       Example: 'middleware' attribute.

    const url = formatUrl(settings.url, endpoint, settings.params);

    return fetch(url, config.before(settings))
      .then((res: any) => createResponse(res, settings))
      .then(config.onResolve)
      .catch(config.onReject);
  }

  function create(instanceConfig: InstanceConfig) {
    return createTrae(merge(config, instanceConfig));
  }

  function get(endpoint: string, requestConfig: RequestInit = {}) {
    const settings: TraeSettings = merge({}, config, requestConfig); 
    return request(endpoint, { ...settings, method: 'GET' });
  }

  function remove(endpoint: string, requestConfig: RequestInit = {}) {
    const settings: TraeSettings = merge({}, config, requestConfig); 
    return request(endpoint, { ...settings, method: 'DELETE' });
  }

  function head(endpoint: string, requestConfig: RequestInit = {}) {
    const settings: TraeSettings = merge({}, config, requestConfig); 
    return request(endpoint, { ...settings, method: 'HEAD' });
  }

  function post(
    endpoint: string,
    data: any = {},
    requestConfig: RequestInit = {},
  ) {
    const settings: TraeSettings = merge({}, config, requestConfig); 
    const body = createRequestBody(data, settings);
    return request(endpoint, { ...settings, method: 'POST', body });
  }

  function put(
    endpoint: string,
    data: any = {},
    requestConfig: RequestInit = {},
  ) {
    const settings: TraeSettings = merge({}, config, requestConfig);
    const body = createRequestBody(data, settings);
    return request(endpoint, { ...settings, method: 'PUT', body });
  }

  function patch(
    endpoint: string,
    data: any = {},
    requestConfig: RequestInit = {},
  ) {
    const settings: TraeSettings = merge({}, config, requestConfig);
    const body = createRequestBody(data, settings);
    return request(endpoint, { ...settings, method: 'PATCH', body });
  }

  return {
    create,
    get,
    delete: remove,
    head,
    post,
    put,
    patch,
    config,
  };
}

export default createTrae();
