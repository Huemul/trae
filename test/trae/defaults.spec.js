/* global describe it expect */

import trae from '../../lib';

describe('trae -> defaults', () => {
  it('returns the current default config when no params are passed', () => {
    expect(trae.defaults()).toMatchSnapshot();
  });

  it('sets the default config to be used on all requests for the instance', () => {
    trae.defaults({ mode: 'no-cors', credentials: 'same-origin' });
    expect(trae.defaults()).toMatchSnapshot();
  });

  it('adds the baseUrl to trae._baseUrl but does not add it to the defaults', () => {
    const apiFoo = trae.create();
    apiFoo.defaults({ baseUrl: '/api/foo' });
    expect(apiFoo._baseUrl).toBe('/api/foo');
    expect(apiFoo._config.get().baseUrl).not.toBeDefined();
  });

  it('adds the baseUrl to trae._baseUrl and add it to the default response', () => {
    const apiFoo = trae.create();
    apiFoo.defaults({ baseUrl: '/api/foo' });
    expect(apiFoo._baseUrl).toBe('/api/foo');
    expect(apiFoo.defaults().baseUrl).toBeDefined();
  });
});
