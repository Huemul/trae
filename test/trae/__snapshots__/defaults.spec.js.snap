// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trae -> defaults returns the current default config when no params are passed 1`] = `
Object {
  "headers": Object {},
  "patch": Object {
    "headers": Object {
      "Content-Type": "application/json",
    },
  },
  "post": Object {
    "headers": Object {
      "Content-Type": "application/json",
    },
  },
  "put": Object {
    "headers": Object {
      "Content-Type": "application/json",
    },
  },
}
`;

exports[`trae -> defaults sets the default config to be used on all requests for the instance 1`] = `
Object {
  "credentials": "same-origin",
  "headers": Object {},
  "mode": "no-cors",
  "patch": Object {
    "headers": Object {
      "Content-Type": "application/json",
    },
  },
  "post": Object {
    "headers": Object {
      "Content-Type": "application/json",
    },
  },
  "put": Object {
    "headers": Object {
      "Content-Type": "application/json",
    },
  },
}
`;
